### ISSUE TITLE
master/slave GDScript keyword name choice might be unfortunate

### ISSUE THREAD

**Issue by TazeTSchnitzel** [godotengine/godot#7986]
There's been something of a trend towards moving away from “master” and “slave” terminology, because of its associations with, well, slavery: https://en.wikipedia.org/wiki/Master-slave_%28technology%29#Appropriateness_of_usage

Besides that issue, sometimes alternative naming choices are clearer anyway (e.g. “leader/follower”), especially to people from a non-technical background. It's not entirely clear to me from the names what they mean. My first guess was it means an authoritative server (“master”) and replicating client (“slave”) relationship, but I'm not quite sure.

At the risk of starting a flame war, it might be worth changing them. If they were to be changed, better to do it now while they're not in a stable release.

**Comment by eon-s** [comment]
I guess it could be a chance to step away from that old nomenclature.

It could be something clear like host-guest, server-client.

**Comment by RandomShaper** [comment]
I think people is too sensitive.

But anyway if the industry is going towards changing the naming, I think Godot should follow.

Do you know any relevant project that has donde the switch? What words have they chosen?

**Comment by TazeTSchnitzel** [comment]
The Wikipedia article I linked to cites a few projects using “replica”.

**Comment by RandomShaper** [comment]
My fault. I didn't read the full article.

I think it would be good to add the meaning of master and slave in Godot to
the first post.

'Replica' seems to be suited to CMS so context is key.


El 9 mar. 2017 1:01 p. m., "Andrea Faulds" <notifications@github.com>
escribió:

The Wikipedia article I linked to cites a few projects using “replica”.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/godotengine/godot/issues/7986#issuecomment-285333394>,
or mute the thread

**Comment by reduz** [comment]
Sorry, I think this is stupid, slavery has been erradicated centuries ago
and there is no one alive that could be offended or discriminated by such
condition.

I understand there are a few retrograde countries still around, but i think
naming convention in a game engine is the least of their problems.

I'm all for stuff like inclusion, equality, human rights, etc. But using
less clear and accepted technical terminology for groundless cause (IMO) is
not a good idea.

**Comment by freemanfromgodotengine** [comment]
Imho if terms master/slave accurately describe exact relation of dependency between two things in Godot, the terms should be used, because they are proper. If that relation between two Godot components is different than master/slave relation, then lets describe it more accurately by using different, proper words. 

In the past, terms like master/slave were used on purpose and not because tech people supported slavery, nor forgot that "host", "guest", "server", "client", "leader", follower", or "replica" words existed. For example, in the context of different audio devices and their synchronization, master/slave relation is really the most accurate term describing how the process works, even to people from a non-music-technical background.

I am against usage of inaccurate Wikipedia entries as a proof of supposed trends or issues, because they may be exaggerated or not true at all.

I checked. Original text by Global Language Monitor says clearly, that it's all about Political (in)Correctness. So, imho, if we are not making politics here with Godot - and we are clearly not, because it's a game engine for all - we should stay with the most logical terms and meanings of words, unless we really want to start creating semantic mess, that we all can observe in the world of contemporary politics - field very often politically correct, but logically deceived.

My $0.02.

**Comment by akien-mga** [comment]
> Sorry, I think this is stupid, slavery has been erradicated centuries ago
and there is no one alive that could be offended or discriminated by such
condition.

Though that's not fully accurate outside of the occidental world, I agree that this is a non-issue. The usage of the terms *master* and *slave* in networking APIs are an industry standard, just like *male* and *female* are standard when talking about electrical sockets (in a pretty graphic way, but at least that's easy to understand).

The terms *master* and *slave* are used to describe the relation between methods and objects in the networking API. There is no personification made at any stage, and it is crystal clear that those concepts have no relation to human slavery. Of course we could find other terms for this, like *capitalist shareholder*/*wage-worker*, but unless there is a clear trend in the industry to move to more politically correct terms, I don't see any gain for Godot Engine to diverge from the standards. It would just make those concepts harder to understand and to link to the wealth of documentation about master/slave networking paradigms.

**Comment by hubbyist** [comment]
I try to be politically correct while talking to students but not with adults and even words are forgotten history do not change. So inventing terminology for political reasons will just cause useless debates IMO.

But it is fun, can we call them narrator/dreamer for making concept more matrix like. :smile:

**Comment by RandomShaper** [comment]
If only a subset of the industry has changed the terms, the standard is still _master_-_slave_. In that case, Godot should keep away from the infection.

**Comment by RandomShaper** [comment]
Sorry for the double comment, but I've just come with the idea that given the fast growth of Godot, soon will come the time when the industry standard will be set by it. :)

**Comment by akien-mga** [comment]
So, case closed I guess.

**Comment by Grass-H0PEr** [comment]
But here's the thing to think about : 

**Children will use this program.**

If it seems harmless to you,
understand that the more exposure a brain has to a phrase,
the more it's repeated internally & propagated & reused.

It's not about liberals or political correctness.

It's about what ideas you want bouncing around in the minds of your children.

**Comment by Grass-H0PEr** [comment]
Progress only happens
when enough people do the right thing.

**Comment by akien-mga** [comment]
Thanks for your input, but the case is closed already.

**Comment by reduz** [comment]
I'm just letting know, almost a year after this issue was closed, and after this was discussed by most of the core contributors, it was decided that an alternative will be added to the keyword "slave" for those who do not wish to be forced use it. 

I hope you understand that, even though the majority was in favor of this change, most of Godot contributors are not from regions where the "slave" word (still) has a strong negative connotation (case in point, most contributors are not from the US), or reside in countries which have significantly more serious and pertinent current social issues (many do not reside in first world countries, have more pressing problems and may find this discussion silly). The world is more complex than many believe it to be. 

This cause took a significant effort to empathize with for many, which yet decided to push for it.. so I hope this effort can be respected and understood.

The "slave" keyword will not be removed for now for compatibility reasons, but the idea is to eventually deprecate it and effectively remove it next time there is a major (compatibility breaking) release. This issue will not be reopened either.

**Comment by reduz** [comment]
As for the words used, we had a long discussion about words to replace it. The main problem is that the traditional words used for this, which are primary/replica, controller/controlled, leader/follower do not really make much sense in the context of how networking in Godot works (because the controlled peer still has to do logic of it's own) so they can be confusing to users (this matters too).

Some discarded proposals included brainslug/human or gru/minion (seriously). We actually did consider keeping the word _master_ as by itself has more positive than negative connotations. We tried  master/minion, master/lackey or master/worker but that still would not have cleared up enough the reference to slavery, which is what can be considered offensive.

Instead, we settled for the term master/puppet, which references to the act of puppetmastering.. which is an activity mainly aimed at entertaining children, with no relation to slavery (or any other abusive relation between humans).

This works better in the Godot context, as the master node will tell the puppet nodes (in different peers) how to appear and act to other players.
